File: Behavioral/Lab18/C#/PriceAdapter.cs
using System;
using System.Collections.Generic;
using System.Text;
using C_.Entities;
using C_.Interfaces;

namespace C_
{
    public class PriceAdapter : IBuyable
    {
        private RealEstate _estate;

        public PriceAdapter(RealEstate estate){
            _estate = estate;
        }

        public void SetEstateToBuy(Customer customer, double price){
            price = ConvertToUSD(customer.Location, price);
            _estate.SetEstateToBuy(customer, price);
        }

        public double ConvertToUSD(string location, double price) => location.ToLower() switch{
            "usa" => price,
            "ukraine" => price / 41.06,
            "france" => price / 841.21,
            "germany" => price / 0.9,
            _ => price
        };
    }
}

File: Structural/Lab9/Typescript/src/price_adapter.ts
import { IBuyable } from './Interfaces/buyable';
import { RealEstate } from './Entities/real_estate';
import { Customer } from './Entities/customer';

export class PriceAdapter implements IBuyable {
    private _estate: RealEstate;

    constructor(estate: RealEstate) {
        this._estate = estate;
    }

    public setEstateToBuy(customer: Customer, price: number): void {
        price = this.convertToUSD(customer.location, price);
        this._estate.setEstateToBuy(customer, price);
    }

    private convertToUSD(location: string, price: number): number {
        switch (location.toLowerCase()) {
            case "usa":
                return price;
            case "ukraine":
                return price / 41.06;
            case "france":
                return price / 841.21;
            case "germany":
                return price / 0.9;
            default:
                return price;
        }
    }
}